### PR TITLE
[Lake] add tablet manager list meta interface

### PR DESCRIPTION
**Pull request by harui7890**
## What type of PR is this：
- [ ] bug
- [X] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

### CODE REVIEW

File: be/src/storage/lake/metadata_iterator.h [hunk 1]
```diff
@@ -1,12 +1,44 @@
 // This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
 
 #pragma once
+#include <vector>
+
+#include "common/status.h"
+#include "common/utils.h"
+#include "fmt/format.h"
```
Comment (imay) [inline be/src/storage/lake/metadata_iterator.h:8]: should use <> for thirdparty include

File: be/src/storage/lake/metadata_iterator.h [hunk 2]
```diff
@@ -1,12 +1,44 @@
...
+    };
+    StatusOr<T> next() {
+        std::string name;
+        if (_pos < _vec.size()) {
+            name = std::move(_vec[_pos++]);
+        } else {
+            return Status::RuntimeError(fmt::format("Out of range pos {} size {}", _pos, _vec.size()));
+        }
+        return get_meta_from_tabletmanager(_group, name);
+    };
```
Comment (sduzh) [inline be/src/storage/lake/metadata_iterator.h:35]: Place these definitions in cpp file and use forward declaration for TabletManager in the header file.

File: be/src/storage/lake/metadata_iterator.h [hunk 3]
```diff
@@ -1,12 +1,44 @@
...
-};
+    explicit MetadataIterator(TabletManager* manager, std::string group, std::vector<std::string> vec)
+            : _manager(manager), _group(std::move(group)), _vec(std::move(vec)), _pos(0){};
+    bool has_next() {
+        if (_pos < _vec.size()) {
+            return true;
+        } else {
+            return false;
+        }
+    };
```
Comment (sduzh) [inline be/src/storage/lake/metadata_iterator.h:26]:
```suggestion
    }
```

File: be/src/storage/lake/metadata_iterator.h (hunk 2 above)
Comment (sduzh) [inline be/src/storage/lake/metadata_iterator.h:35]:
```suggestion
    }
```

File: be/src/storage/lake/metadata_iterator.h [hunk 4]
```diff
@@ -1,12 +1,44 @@
...
+        if (_pos < _vec.size()) {
+            name = std::move(_vec[_pos++]);
+        } else {
+            return Status::RuntimeError(fmt::format("Out of range pos {} size {}", _pos, _vec.size()));
+        }
+        return get_meta_from_tabletmanager(_group, name);
+    };
 
+private:
+    StatusOr<T> get_meta_from_tabletmanager(const std::string& group, const std::string& path);
```
Comment (sduzh) [inline be/src/storage/lake/metadata_iterator.h:38]:
```suggestion
    StatusOr<T> get_meta_from_tabletmanager(const std::string& group, const std::string& path);
    
```

File: be/src/storage/lake/metadata_iterator.h [hunk 5]
```diff
@@ -1,12 +1,44 @@
...
+#include <vector>
+
+#include "common/status.h"
+#include "common/utils.h"
+#include "fmt/format.h"
+#include "storage/lake/tablet_manager.h"
 
 namespace starrocks::lake {
 
-// TODO:
```
Comment (sduzh) [inline be/src/storage/lake/metadata_iterator.h:7]: What does this TODO for?

File: be/src/storage/lake/metadata_iterator.h [hunk 6]
```diff
@@ -1,12 +1,44 @@
...
+class TabletManager;
```
Comment (sduzh) [inline be/src/storage/lake/metadata_iterator.h:13]: Unused

File: be/src/storage/lake/metadata_iterator.h [hunk 7]
```diff
@@ -1,12 +1,44 @@
 // This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
 
 #pragma once
+#include <vector>
+
+#include "common/status.h"
+#include "common/utils.h"
+#include "fmt/format.h"
+#include "storage/lake/tablet_manager.h"
```
Comment (sduzh) [inline be/src/storage/lake/metadata_iterator.h:9]: The include is unnecessary, you can remove it.

Comment (sduzh) [inline be/src/storage/lake/metadata_iterator.h:9]: The include is unnecessary, you can remove it.

File: be/src/storage/lake/metadata_iterator.h [hunk 8]
```diff
@@ -1,12 +1,44 @@
 // This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
 
 #pragma once
+#include <vector>
```
Comment (sduzh) [inline be/src/storage/lake/metadata_iterator.h:4]:
```suggestion

#include <vector>
```

File: be/src/storage/lake/metadata_iterator.cpp [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+#include "storage/lake/txn_log.h"
+
+namespace starrocks::lake {
+class TabletManager;
+
+template <>
+StatusOr<TabletMetadataPtr> MetadataIterator<TabletMetadataPtr>::get_meta_from_tabletmanager(const std::string& group,
+                                                                                             const std::string& path) {
+    return _manager->get_tablet_metadata(group, path);
+};
```
Comment (sduzh) [inline be/src/storage/lake/metadata_iterator.cpp:15]:
```suggestion
}

```

Comment (sduzh) [reply be/src/storage/lake/metadata_iterator.cpp:15]: Remove ';' and leave a blank line between the next function.

File: be/src/storage/lake/metadata_iterator.cpp [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+template <>
+StatusOr<TxnLogPtr> MetadataIterator<TxnLogPtr>::get_meta_from_tabletmanager(const std::string& group,
+                                                                             const std::string& path) {
+    return _manager->get_txn_log(group, path);
+};
```
Comment (sduzh) [inline be/src/storage/lake/metadata_iterator.cpp:20]:
```suggestion
}
```

File: be/src/storage/lake/tablet_manager.cpp [hunk 11]
```diff
@@ -116,32 +128,56 @@ StatusOr<TabletMetadataPtr> TabletManager::get_tablet_metadata(const std::string
...
+        prefix = fmt::format("tbl_{:016X}_", tablet_id);
+    }
+    ASSIGN_OR_RETURN(auto fs, FileSystem::CreateSharedFromString(group));
+    auto scan_cb = [&objects, &prefix](std::string_view name) {
+        if (!HasPrefixString(name, prefix)) {
+            // skip object for other tablet
+            return true;
+        }
+        objects.emplace_back(name);
+        return true;
```
Comment (sduzh) [inline be/src/storage/lake/tablet_manager.cpp:172]:
```suggestion
        if (HasPrefixString(name, prefix)) {
                objects.emplace_back(name);
        }
        return true;
```

File: be/src/storage/lake/metadata_iterator.cpp [hunk 12]
```diff
@@ -0,0 +1,22 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+#include "storage/lake/metadata_iterator.h"
+
+#include "storage/lake/tablet_manager.h"
+#include "storage/lake/tablet_metadata.h"
+#include "storage/lake/txn_log.h"
+
+namespace starrocks::lake {
+class TabletManager;
```
Comment (sduzh) [inline be/src/storage/lake/metadata_iterator.cpp:9]: This forward declaration is unnecessary.

File: be/src/storage/lake/metadata_iterator.h [hunk 13]
```diff
@@ -1,12 +1,44 @@
...
 public:
-};
+    explicit MetadataIterator(TabletManager* manager, std::string group, std::vector<std::string> vec)
+            : _manager(manager), _group(std::move(group)), _vec(std::move(vec)), _pos(0){};
+    bool has_next() {
+        if (_pos < _vec.size()) {
+            return true;
+        } else {
+            return false;
+        }
```
Comment (sduzh) [inline be/src/storage/lake/metadata_iterator.h:25]:
```suggestion
        return _pos < _vec.size();
```

File: be/src/storage/lake/metadata_iterator.h (hunk 4 above)
Comment (sduzh) [reply be/src/storage/lake/metadata_iterator.h:38]: Leave a blank line between the method and variables.

File: be/src/storage/lake/metadata_iterator.h [hunk 14]
```diff
@@ -1,12 +1,44 @@
...
+    };
+    StatusOr<T> next() {
+        std::string name;
+        if (_pos < _vec.size()) {
+            name = std::move(_vec[_pos++]);
+        } else {
+            return Status::RuntimeError(fmt::format("Out of range pos {} size {}", _pos, _vec.size()));
+        }
+        return get_meta_from_tabletmanager(_group, name);
```
Comment (sduzh) [inline be/src/storage/lake/metadata_iterator.h:34]:
```suggestion
        if (_pos < _vec.size()) {
            return get_meta_from_tabletmanager(_group, _vec[_pos++]);
        } else {
            return Status::RuntimeError(fmt::format("Out of range pos {} size {}", _pos, _vec.size()));
        }
        
```

File: be/src/storage/lake/tablet_manager.cpp [hunk 15]
```diff
@@ -157,31 +193,55 @@ StatusOr<TxnLogPtr> TabletManager::get_txn_log(const std::string& group, int64_t
...
+    }
+    // get group path
+    ASSIGN_OR_RETURN(auto fs, FileSystem::CreateSharedFromString(group));
+    auto scan_cb = [&objects, &prefix](std::string_view name) {
+        if (!HasPrefixString(name, prefix)) {
+            // skip object for other tablet
+            return true;
+        }
+        objects.emplace_back(name);
+        return true;
```
Comment (sduzh) [inline be/src/storage/lake/tablet_manager.cpp:240]:
```suggestion
        if (HasPrefixString(name, prefix)) {
            objects.emplace_back(name);
        }
        return true;
```

File: be/src/storage/lake/tablet_manager.h [hunk 16]
```diff
@@ -43,14 +46,20 @@ class TabletManager {
...
 private:
     std::string tablet_meta_path(const std::string& group, int64_t tablet_id, int64_t verson);
+    std::string tablet_meta_path(const std::string& group, const std::string& meta_path);
     std::string txn_log_path(const std::string& group, int64_t tablet_id, int64_t txn_id);
-    StatusOr<MetadataIterator> list_tablet_metadata(const std::string& group);
-    StatusOr<MetadataIterator> list_tablet_metadata(const std::string& group, int64_t tablet_id);
-    StatusOr<MetadataIterator> list_txn_log(const std::string& group);
-    StatusOr<MetadataIterator> list_txn_log(const std::string& group, int64_t tablet_id);
+    std::string txn_log_path(const std::string& group, const std::string& txnlog_path);
+    StatusOr<TabletMetadataPtr> _get_tablet_metadata(const std::string& meta_full_path);
```
Comment (sduzh) [inline be/src/storage/lake/tablet_manager.h:61]:
```suggestion
    StatusOr<TabletMetadataPtr> get_tablet_metadata(const std::string& meta_full_path);
```
 Keep a consistent code style.

File: be/src/storage/lake/tablet_manager.h [hunk 17]
```diff
@@ -43,14 +46,20 @@ class TabletManager {
...
+    StatusOr<TxnLogPtr> _get_txn_log(const std::string& txnlog_full_path);
```
Comment (sduzh) [inline be/src/storage/lake/tablet_manager.h:62]:
```suggestion
    StatusOr<TxnLogPtr> get_txn_log(const std::string& txnlog_full_path);
```

File: be/src/storage/lake/metadata_iterator.h [hunk 18]
```diff
@@ -2,11 +2,41 @@
 
 #pragma once
 
+#include <vector>
+
+#include "common/status.h"
+#include "fmt/format.h"
```
Comment (sduzh) [inline be/src/storage/lake/metadata_iterator.h:8]:
```suggestion
#include <fmt/format.h>
```
```suggestion
#include "fmt/format.h"
```

Comment (sduzh) [inline be/src/storage/lake/metadata_iterator.h:8]:
```suggestion
#include <fmt/format.h>
```
```suggestion
#include "fmt/format.h"
```

File: be/src/storage/lake/metadata_iterator.h [hunk 19]
```diff
@@ -2,11 +2,41 @@
...
-};
+    explicit MetadataIterator(TabletManager* manager, std::string group, std::vector<std::string> vec)
+            : _manager(manager), _group(std::move(group)), _vec(std::move(vec)), _pos(0){};
 
+    bool has_next() {
+        return _pos < _vec.size();
+    }
+
+    StatusOr<T> next() {
+        std::string name;
```
Comment (sduzh) [inline be/src/storage/lake/metadata_iterator.h:26]: unused variable

Comment (sduzh) [inline be/src/storage/lake/metadata_iterator.h:26]: unused variable

File: be/src/storage/lake/metadata_iterator.h (hunk 18 above)
Comment (sduzh) [inline be/src/storage/lake/metadata_iterator.h:8]:
```suggestion
#include <fmt/format.h>
```

**Review by sduzh — DISMISSED** [review]

File: be/src/storage/lake/metadata_iterator.h [hunk 20]
```diff
@@ -2,11 +2,39 @@
...
+    StatusOr<T> next() {
+        if (_pos < _vec.size()) {
+            return get_meta_from_tabletmanager(_group, _vec[_pos++]);
+        } else {
+            return Status::RuntimeError(fmt::format("Out of range pos {} size {}", _pos, _vec.size()));
+        }
+    }
+
+private:
+    StatusOr<T> get_meta_from_tabletmanager(const std::string& group, const std::string& path);
```
Comment (wyb) [inline be/src/storage/lake/metadata_iterator.h:33]:
```suggestion
    StatusOr<T> get_meta_from_tablet_manager(const std::string& group, const std::string& path);
```

File: be/src/storage/lake/metadata_iterator.h [hunk 21]
```diff
@@ -2,11 +2,39 @@
...
 
-// TODO:
+class TabletManager;
+
+template <typename T>
 class MetadataIterator {
 public:
-};
+    explicit MetadataIterator(TabletManager* manager, std::string group, std::vector<std::string> vec)
+            : _manager(manager), _group(std::move(group)), _vec(std::move(vec)), _pos(0){};
```
Comment (wyb) [inline be/src/storage/lake/metadata_iterator.h:20]:
```suggestion
            : _manager(manager), _group(std::move(group)), _vec(std::move(vec)), _pos(0) {};
```

File: be/src/storage/lake/tablet_manager.cpp [hunk 22]
```diff
@@ -29,6 +29,14 @@ std::string TabletManager::tablet_meta_path(const std::string& group, int64_t ta
     }
 }
 
+std::string TabletManager::tablet_meta_path(const std::string& group, const std::string& meta_path) {
```
Comment (wyb) [inline be/src/storage/lake/tablet_manager.cpp:32]:
```suggestion
std::string TabletManager::tablet_metadata_path(const std::string& group, const std::string& metadata_path) {
```

**Review by wyb — APPROVED** [review]

**Review by sduzh — APPROVED** [review]

**Pull request merged**